sdk/inc/azure/core/az_mqtt5_rpc_client.h

```diff
@@ -0,0 +1,334 @@
+    _az_hfsm rpc_client_policy;
+
+    /**
+     * @brief The subclient used by the MQTT5 RPC Client.
+     */
+    _az_event_client subclient;
+
+    /**
+     * @brief The MQTT5 connection linked to the MQTT5 RPC Client.
+     */
+    az_mqtt5_connection* connection;
+
+    /**
+     * @brief the message id of the pending subscribe for the response topic
+     */
+    int32_t pending_subscription_id;
+
+    /**
+     * @brief timer used for the subscribe of the response topic
+     */
+    _az_event_pipeline_timer rpc_client_timer;
+
+    /**
+     * @brief az_mqtt5_rpc_client associated with this hfsm
+     */
+    az_mqtt5_rpc_client* rpc_client;
+
+    /**
+     * @brief The property bag used by the RPC client policy for sending request messages
+     */
+    az_mqtt5_property_bag property_bag;
+
+  } _internal;
+};
+
+// Event data types
+
+/**
+ * @brief Event data for #AZ_EVENT_RPC_CLIENT_INVOKE_REQ.
+ */
+typedef struct az_mqtt5_rpc_client_invoke_req_event_data
+{
+  /**
+   * @brief The correlation id of the request.
+   */
+  az_span correlation_id;
+
+  /**
+   * @brief The content type of the request.
+   */
+  az_span content_type;
+
+  /**
+   * @brief The payload of the request.
+   */
+  az_span request_payload;
+
+  /**
+   * @brief The message id of the request to correlate with pubacks.
+   */
+  int32_t mid;
+} az_mqtt5_rpc_client_invoke_req_event_data;
+
+/**
+ * @brief Event data for #AZ_EVENT_RPC_CLIENT_RSP.
+ */
+typedef struct az_mqtt5_rpc_client_rsp_event_data
+{
+  /**
+   * @brief The correlation id of the response.
+   */
+  az_span correlation_id;
+
+  /**
+   * @brief The error message of the response.
+   */
+  az_span error_message;
+
+  /**
+   * @brief The status of the response.
+   */
+  az_mqtt5_rpc_status status;
+
+  /**
+   * @brief The content type of the response.
+   */
+  az_span content_type;
+
+  /**
+   * @brief The payload of the response.
+   */
+  az_span response_payload;
+
+  /**
+   * @brief Whether the RPC client failed to parse the response - if the correlation id was able to
+   * be parsed, the request can still be considered completed if desired.
+   */
+  bool parsing_failure;
+} az_mqtt5_rpc_client_rsp_event_data;
+
+/**
+ * @brief Triggers an AZ_EVENT_RPC_CLIENT_INVOKE_REQ event from the application
+ *
+ * @note This should be called from the application when wants to request a command to be invoked.
+ *
+ * @param[in] client The az_mqtt5_rpc_client_hfsm to use.
+ * @param[in] data The information for the execution request
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK The event was triggered successfully.
+ * @retval #AZ_ERROR_HFSM_INVALID_STATE If called when the HFSM hasn't been asked to subscribe yet
+ * or is still subscribing.
+ * @retval #AZ_ERROR_NOT_SUPPORTED if the client is not connected.
+ * @retval Other on other failures creating/sending the request message.
+ */
+AZ_NODISCARD az_result az_mqtt5_rpc_client_invoke_req(
```

Comment:
I recommend using the required [naming conventions for long-running operations](https://azure.github.io/azure-sdk/clang_design.html#:~:text=DO%20use%20az_%3Cshortname%3E_%3Cobjname%3E_begin_%3Cverb%3E_%3Cnoun%3E%20for%20methods%20that%20initiate%20a%20long%20running%20operation.%20For%20example%3A%20az_storage_blob_begin_copy_from_url().) 

`az_mqtt5_rpc_client_invoke_begin()`

Also, all long-running operations require an `az_context` that can be cancelled (even if for now, we won't do much with it).
Please pass the az_context into az_mqtt: it should be verified prior to forwarding data to the MQTT stack.

Reply:
Will add the naming change.
For the `begin` naming and az_context on long running operations, I didn't see any other example of this in the SDK to base it off of. I could definitely add an az_context for these requests so that it can be cancelled until the MQTT send is triggered, but right now we don't have this functionality in az_mqtt to accept a context that could be cancelled. I'm going to hold off on adding this feature to az_mqtt for now (and therefore the az_context on invoke/sub/unsub requests), but I'll keep track of it as a future feature